### PR TITLE
Add Anticaptcha support for solving captcha

### LINKED ISSUE #55: Captcha?
How can we get this to work with the captcha?

Kr

### PR DESCRIPTION
**Pull request by ad-m**
Fix #55, #51 using captcha solving service. My referral link to this service: http://getcaptchasolution.com/ewfhjk64ll  The captcha cost is avg. 0.0022 $ / each (2$ / 1k captcha). There is no minimal deposit. So for a small amount you can have a service that works for years.

Naprawia #55, #51 z użyciem usługi rozwiązywania kodów obrazkowych. Mój link polecający: http://getcaptchasolution.com/ewfhjk64ll Koszt za jeden obrazek to 0.0022 $ / każdy (2 $ / 1 k). Nie ma minimalnego depozytu. Za drobne można mieć usługę działającą przez lata.

### CODE REVIEW

**Comment by ad-m** [comment]
Example log:
```
[INFO] - Creating session...
[INFO] - Session created, logged in successfully!
[INFO] - Start grabbing eBook...
[INFO] - Captcha detected. Trying to solve it using Anti-captcha.com. 
[INFO] - Task ID 211173825 created. Waiting to finish.
[INFO] - Taks finished in 21.98 second
[SUCCESS] - eBook: 'Learning_Python' has been successfully grabbed!
[SUCCESS] - Good, looks like all went well! :-)
```

**Comment by igbt6** [comment]
Hej Adam, seems like a nice try !  I'm on the vacation till the end of the week, but definetely test  your code and merge if ok somehow close to Sunday, Best regards!

**Comment by ad-m** [comment]
I made changes. I've separated Anticaptcha support into a separate library, as it can work independently, which can provide more users and more efficient development. 

There is a risk that Google will improve Recaptcha security, and additional actions will be needed, such as creating a proxy for one IP address for captcha solving and validating the form.

**Comment by johnpark82** [comment]
I've been waiting for this feature :)

**Comment by ad-m** [comment]
@johnpark82 , have you tested that? Do you have any issues?

**Comment by NemoKOF** [comment]
I'm using this since release, it works well. Although this weekend I got 2 failed attempts in grabbing books. I think, they are not published on weekends but I'm not sure if it's that. The script runs everyday started by anacron. Log:

`2017-06-30 07:36:05,954 - __main__ - [SUCCESS] - Code for eBook: 'Mastering_Julia' downloaded successfully!
2017-06-30 07:36:05,954 - __main__ - [INFO] - 4 eBooks have been downloaded!
2017-06-30 07:36:05,954 - __main__ - [SUCCESS] - Good, looks like all went well! :-)
2017-07-01 07:35:09,482 - __main__ - [INFO] - Creating session...
2017-07-01 07:35:16,826 - __main__ - [INFO] - Session created, logged in successfully!
2017-07-01 07:35:16,827 - __main__ - [INFO] - Start grabbing eBook...
2017-07-01 07:35:17,201 - __main__ - [ERROR] - Exception occurred argument of type 'NoneType' is not iterable
2017-07-02 07:35:09,329 - __main__ - [INFO] - Creating session...
2017-07-02 07:35:13,939 - __main__ - [INFO] - Session created, logged in successfully!
2017-07-02 07:35:13,940 - __main__ - [INFO] - Start grabbing eBook...
2017-07-02 07:35:14,191 - __main__ - [ERROR] - Exception occurred argument of type 'NoneType' is not iterable
2017-07-03 07:35:09,196 - __main__ - [INFO] - Creating session...
2017-07-03 07:35:14,438 - __main__ - [INFO] - Session created, logged in successfully!
2017-07-03 07:35:14,438 - __main__ - [INFO] - Start grabbing eBook...
2017-07-03 07:35:14,635 - __main__ - [INFO] - Captcha detected. Trying to solve it using Anti-captcha.com.
2017-07-03 07:35:15,287 - __main__ - [INFO] - Task ID 233812274 created. Waiting to finish.
2017-07-03 07:35:24,857 - __main__ - [INFO] - Taks finished in 9.57 second
2017-07-03 07:35:27,295 - __main__ - [SUCCESS] - eBook: 'Learning_Neo4j' has been successfully grabbed!
2017-07-03 07:35:27,295 - __main__ - [INFO] - Getting data of all your books...
2017-07-03 07:35:29,472 - __main__ - [INFO] - Opened 'https://www.packtpub.com/account/my-ebooks' successfully!
2017-07-03 07:35:30,636 - __main__ - [INFO] - Title: 'Learning_Neo4j'
2017-07-03 07:35:30,639 - __main__ - [INFO] - Downloading eBook: 'Learning_Neo4j' in .pdf format...
2017-07-03 07:35:31,457 - __main__ - [SUCCESS] - eBook: 'Learning_Neo4j.pdf' downloaded successfully!
2017-07-03 07:35:31,457 - __main__ - [INFO] - Title: 'Learning_Neo4j'
2017-07-03 07:35:31,460 - __main__ - [INFO] - Downloading eBook: 'Learning_Neo4j' in .epub format...
2017-07-03 07:35:32,874 - __main__ - [SUCCESS] - eBook: 'Learning_Neo4j.epub' downloaded successfully!
2017-07-03 07:35:32,875 - __main__ - [INFO] - Title: 'Learning_Neo4j'
2017-07-03 07:35:32,875 - __main__ - [INFO] - Downloading eBook: 'Learning_Neo4j' in .mobi format...
2017-07-03 07:35:34,797 - __main__ - [SUCCESS] - eBook: 'Learning_Neo4j.mobi' downloaded successfully!
2017-07-03 07:35:34,797 - __main__ - [INFO] - 3 eBooks have been downloaded!
2017-07-03 07:35:34,797 - __main__ - [SUCCESS] - Good, looks like all went well! :-)
`

**Comment by ad-m** [comment]
At weeknds I received this exceptions too....

In addiction there is following exception:
```
[INFO] - Creating session...
[INFO] - Session created, logged in successfully!
[INFO] - Start grabbing eBook...
[INFO] - Captcha detected. Trying to solve it using Anti-captcha.com.
[ERROR] - Exception occurred HTTPSConnectionPool(host='api.anti-captcha.com', port=443): Max retries exceeded with url: /createTask (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7f5711883050>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))
```

It's weird. Dig doesn't report this kind of issues. Any others requests on this servers too.

File: src/packtPublishingFreeEbook.py [hunk 1]
```diff
@@ -170,18 +172,36 @@ def get_ebook_infodata(self, r):
         self._write_ebook_infodata(result_data)
         return result_data
 
-    def grab_ebook(self, log_ebook_infodata = False):
+    def solve_captcha(self, url, html):
+        key_pattern = re.compile("Packt.offers.onLoadRecaptcha\(\'(.+?)\'\)")
+        site_key = key_pattern.search(html.find(text=key_pattern)).group(1)
```
Comment (ad-m) [inline src/packtPublishingFreeEbook.py]: Is it possible to write better without double search? How?

**Comment by ad-m** [comment]
Changed DNS server. No more DNS resolve errors.

**Comment by Santosh-Sah** [comment]
I think you can stop your development on anti-CAPTCHA as they have removed it.

**Comment by Santosh-Sah** [comment]
They have removed the CAPTCHA field but still below getting error.
`[INFO] - Creating session...
/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/util/ssl_.py:334: SNIMissingWarning: An HTTPS request has been made, but the SNI (Subject Name Indication) extension to TLS is not available on this platform. This may cause the server to present an incorrect TLS certificate, which can cause validation failures. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  SNIMissingWarning
/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/util/ssl_.py:132: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecurePlatformWarning
/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/util/ssl_.py:132: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecurePlatformWarning
[INFO] - Session created, logged in successfully!
[INFO] - Start grabbing eBook...
[ERROR] - Exception occurred u'href'`

**Comment by igbt6** [comment]
I just wanted to merge it but exactly it seems that Packt has been playing with reCaptcha from yesterday, as for now they have not removed reCaptcha, just changed a mode of using it or sth. Let's wait for some update :>

**Comment by Santosh-Sah** [comment]
Yeah I have inspected and found it they are still using it but kind of stealth way. 

`<div class="grecaptcha-badge" style="width: 256px; height: 60px; transition: right 0.3s ease; position: fixed; bottom: 14px; right: -186px; box-shadow: gray 0px 0px 5px;"><div class="grecaptcha-logo"><iframe src="https://www.google.com/recaptcha/api2/anchor?k=6LeAHSgUAAAAAKsn5jo6RUSTLVxGNYyuvUcLMe0_&amp;co=aHR0cHM6Ly93d3cucGFja3RwdWIuY29tOjQ0Mw..&amp;hl=en&amp;v=r20170712134223&amp;size=invisible&amp;cb=13tk8z5yy5g" title="recaptcha widget" width="256" height="60" frameborder="0" scrolling="no" sandbox="allow-forms allow-modals allow-popups allow-popups-to-escape-sandbox allow-same-origin allow-scripts allow-top-navigation"></iframe></div><div class="grecaptcha-error"></div><textarea id="g-recaptcha-response" name="g-recaptcha-response" class="g-recaptcha-response" style="width: 250px; height: 40px; border: 1px solid #c1c1c1; margin: 10px 25px; padding: 0px; resize: none;  display: none; "></textarea></div>`

**Comment by ad-m** [comment]
This tools bypass captchas for today:

```
[INFO] - Creating session...
[INFO] - Session created, logged in successfully!
[INFO] - Start grabbing eBook...
[INFO] - Captcha detected. Trying to solve it using Anti-captcha.com.
[INFO] - Task ID 56395448 created. Waiting to finish.
[INFO] - Taks finished in 31.67 second
[SUCCESS] - eBook: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook' has been successfully grabbed!
[INFO] - Getting data of all your books...
[INFO] - Opened 'https://www.packtpub.com/account/my-ebooks' successfully!
[INFO] - Title: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook'
[INFO] - Downloading eBook: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook' in .pdf format...
[SUCCESS] - eBook: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook.pdf' downloaded successfully!
[INFO] - Title: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook'
[INFO] - Downloading eBook: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook' in .epub format...
[SUCCESS] - eBook: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook.epub' downloaded successfully!
[INFO] - Title: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook'
[INFO] - Downloading eBook: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook' in .mobi format...
[SUCCESS] - eBook: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook.mobi' downloaded successfully!
[INFO] - Title: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook'
[INFO] - Downloading code for eBook: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook'...
[SUCCESS] - Code for eBook: 'HTML5_and_CSS3_Responsive_Web_Design_Cookbook' downloaded successfully!
[INFO] - 4 eBooks have been downloaded!
[SUCCESS] - Good, looks like all went well! :-)
```

It looks they use https://developers.google.com/recaptcha/docs/invisible .

**Comment by ad-m** [comment]
@igbt6 , to już działa. Ukryli captchę, ale ona wciąż jest, ale kod nie wymagał zmiany. Kod nie działa, gdy nie ma w danym dniu książki. Problem z DNS-ami zostały naprawione, bo to usterka dla mnie.

**Comment by ranrinc** [comment]
Hi all,

would like to implement the anti-captcha feature. Should I go ahead? Is this running good? Should I wait? Let me know

**Comment by ad-m** [comment]
@ranrinc , it's done. It works stable and reliable. Ready to merge.

```
[INFO] - Creating session...
[INFO] - Session created, logged in successfully!
[INFO] - Start grabbing eBook...
[INFO] - Captcha detected. Trying to solve it using Anti-captcha.com.
[INFO] - Task ID 71628955 created. Waiting to finish.
[INFO] - Taks finished in 21.92 second
[SUCCESS] - eBook: 'Arduino_Development_Cookbook' has been successfully grabbed!
[INFO] - Getting data of all your books...
[INFO] - Opened 'https://www.packtpub.com/account/my-ebooks' successfully!
[INFO] - Title: 'Arduino_Development_Cookbook'
[INFO] - Downloading eBook: 'Arduino_Development_Cookbook' in .pdf format...
[SUCCESS] - eBook: 'Arduino_Development_Cookbook.pdf' downloaded successfully!
[INFO] - Title: 'Arduino_Development_Cookbook'
[INFO] - Downloading eBook: 'Arduino_Development_Cookbook' in .epub format...
[SUCCESS] - eBook: 'Arduino_Development_Cookbook.epub' downloaded successfully!
[INFO] - Title: 'Arduino_Development_Cookbook'
[INFO] - Downloading eBook: 'Arduino_Development_Cookbook' in .mobi format...
[SUCCESS] - eBook: 'Arduino_Development_Cookbook.mobi' downloaded successfully!
[INFO] - Title: 'Arduino_Development_Cookbook'
[INFO] - Downloading code for eBook: 'Arduino_Development_Cookbook'...
[SUCCESS] - Code for eBook: 'Arduino_Development_Cookbook' downloaded successfully!
[INFO] - 4 eBooks have been downloaded!
[SUCCESS] - Good, looks like all went well! :-)
```

**Comment by igbt6** [comment]
Yes, Yes, I have also use it but forgot to merge :D  @ad-m Maybe you could add some info to README.md describing how to set it with anticaptcha.com  and some useful stuff. After that i will merge it :) Thx !

**Comment by ad-m** [comment]
@igbt6 , fixed. I used own referral link as the author of a patch. If you want - change to your referral link as the author of the software. It does not extra cost of user, and motivates to provide constant support.

**Comment by igbt6** [comment]
Hello @ad-m  
Finally merged, thx for a nice piece of code ;)

**Comment by Forge64** [comment]
I was excited to get this working, until I actually started setting up an anti-captcha account. Lots and lots of hoops, and then after I tried to add funds to the account, they want a scan of my credit card or a federal ID. No way. Not happening. The site is eighteen kinds of shady, which is to be expected given their line of business, but I'm not scanning my driver's license and sending it to Eastern Europe. That's just a recipe for trouble. It's a shame there's no better solution, but I'll keep solving and downloading by hand for now.

**Comment by ranrinc** [comment]
@Forge64 Try paying it using PayPal, still the safest out there

**Comment by ad-m** [comment]
@Forge64 Anti-captcha support a lot of gateway to add funs. Probably anti-fraud system of gateway detect something suspicious in your transaction, so require extra verification. Just change gateway and do not upload any documents. 

I use the service without presenting any documents. Anti-captcha service in most countries does not violate the law in any way. Another situation is some platform users who can - like us - exploit them contrary to the regulations of electronic service providers.

**Comment by Forge64** [comment]
I had tried to use a credit card directly. I may try Paypal later, but I dislike doing business with them as well. It also requires a minimum 10$ deposit via Paypal vs. 1$ via CC direct. Since I'm using this only for packt, and only twice a day, 10$ of credit is something like 5 years.

**Comment by Forge64** [comment]
Paypal is not on the list of payment options for me. Most seem to be in Cyrillic. The only options I recognize are Visa/MC, which have the aforementioned problem. This solution doesn't work for me. Shame on Packt.

**Pull request merged**